braindecode/preprocessing/preprocess.py

```diff
@@ -130,34 +144,73 @@ def preprocess(concat_ds, preprocessors):
         assert hasattr(elem, 'apply'), (
             'Preprocessor object needs an `apply` method.')
 
-    for ds in concat_ds.datasets:
-        if hasattr(ds, 'raw'):
-            _preprocess(ds.raw, preprocessors)
-        elif hasattr(ds, 'windows'):
-            _preprocess(ds.windows, preprocessors)
-        else:
-            raise ValueError(
-                'Can only preprocess concatenation of BaseDataset or '
-                'WindowsDataset, with either a `raw` or `windows` attribute.')
+    list_of_ds = Parallel(n_jobs=n_jobs)(
+        delayed(_preprocess)(ds, i, preprocessors, save_dir, overwrite)
+        for i, ds in enumerate(concat_ds.datasets))
 
-    # Recompute cumulative sizes as the transforms might have changed them
-    # XXX: Ultimately, the best solution would be to have cumulative_size be
-    #      a property of BaseConcatDataset.
-    concat_ds.cumulative_sizes = concat_ds.cumsum(concat_ds.datasets)
+    if save_dir is not None:  # Reload datasets and replace in concat_ds
+        concat_ds_reloaded = load_concat_dataset(
+            save_dir, preload=False, target_name=None)
+        _replace_inplace(concat_ds, concat_ds_reloaded)
+    else:
+        if n_jobs is None or n_jobs == 1:  # joblib did not make copies, the
+            # preprocessing happened in-place
+            # Recompute cumulative sizes as transforms might have changed them
+            concat_ds.cumulative_sizes = concat_ds.cumsum(concat_ds.datasets)
+        else:  # joblib made copies
+            _replace_inplace(concat_ds, BaseConcatDataset(list_of_ds))
 
 
-def _preprocess(raw_or_epochs, preprocessors):
+def _replace_inplace(concat_ds, new_concat_ds):
+    if len(concat_ds.datasets) != len(new_concat_ds.datasets):
+        raise ValueError('Both inputs must have the same length.')
+    for i in range(len(new_concat_ds.datasets)):
+        concat_ds.datasets[i] = new_concat_ds.datasets[i]
+
+
+def _preprocess(ds, ds_index, preprocessors, save_dir=None, overwrite=False):
     """Apply preprocessor(s) to Raw or Epochs object.
 
     Parameters
     ----------
-    raw_or_epochs: mne.io.Raw or mne.Epochs
-        Object to preprocess.
+    ds: BaseDataset
+        Dataset object to preprocess.
+    ds_index : int
+        Index of the BaseDataset in its BaseConcatDataset. Ignored if save_dir
+        is None.
     preprocessors: list(Preprocessor)
         List of preprocessors to apply to the dataset.
+    save_dir : str | None
+        If provided, save the preprocessed BaseDataset in the
+        specified directory.
+    overwrite : bool
+        If True, overwrite existing file with the same name.
```

Comment:
Will delete evverything, not just the file with the same name.

Reply:
When you say everything, you mean `description.json` and `target_name.json`, right?

Reply:
Overwriting will be on subdirectory level